Sort nodes in ascending order for HighNodeUtilization

dineshbhor opened the pull request:
Sort nodes in ascending order for HighNodeUtilization. 
This fixes the issue reported in https://github.com/kubernetes-sigs/descheduler/issues/756 and a part of issue in https://github.com/kubernetes-sigs/descheduler/issues/725

JaneLiuL commented:
hi @dineshbhor  thanks for bring this

dineshbhor commented:
@JaneLiuL Thanks for the comment. I kept the logic the same as LowNodeUtilization, just in ascending order.
As the discussion proceeds the approach is still not decided so for the time being until we finalize the approach we can keep the same behavior as LowNodeUtilization just the order is ascending.

damemi reviewed (commented):
/ok-to-test

dineshbhor commented:
Hello @JaneLiuL, @damemi
Thanks for taking the time to review this PR.
We have 2 GKE clusters (dev-670, prod-1200 approx) for which we are trying to introduce a descheduler along with an inbuild managed cluster autoscaler to reduce the overall cost (`HighNodeUtilization`).

This bug is a kind of blocker for that work. If possible it would be really helpful to get this PR reviewed/merged/released sooner.

Thanks in advance 🙇 
Dinesh

ingvagabund commented on pkg/descheduler/strategies/nodeutilization/nodeutilization.go:
```diff
@@ -213,7 +213,7 @@ func evictPodsFromSourceNodes(
 ) {
 
-	sortNodesByUsage(sourceNodes)
+	sortNodesByUsage(sourceNodes, strategy)
```
Instead of passing the strategy name into `sortNodesByUsage`, would you mind changing the signature to `func sortNodesByUsage(nodes []NodeInfo, ascending bool)` and moving `sortNodesByUsage` function invocation from `evictPodsFromSourceNodes`? I.e.
```go
# LowNodeUtilization
sortNodesByUsage(sourceNodes, false)
evictPodsFromSourceNodes(
		ctx,
		sourceNodes,
		lowNodes,
		podEvictor,
		evictable.IsEvictable,
		resourceNames,
		"LowNodeUtilization",
		continueEvictionCond)
```
and
```go
# HighNodeUtilization
sortNodesByUsage(sourceNodes, true)
evictPodsFromSourceNodes(
		ctx,
		sourceNodes,
		highNodes,
		podEvictor,
		evictable.IsEvictable,
		resourceNames,
		"HighNodeUtilization",
		continueEvictionCond)
```

damemi reviewed (commented):
Please add a config setting for this so it isn't turned on by default, and also test cases to cover it. Thanks!

dineshbhor commented:
Thanks, @ingvagabund for the comment. Fixed it.

@damemi
> Please add a config setting for this so it isn't turned on by default, and also test cases to cover it. Thanks!
#### Config setting for this so it isn't turned on by default
Correct me if I am wrong.
This was a bug in existing behavior which will be fixed by this PR. So there is no change in the behavior from the user's perspective. So in my opinion there is no need to introduce a config setting. If I misunderstood your request please share more information.

ingvagabund commented:
Agree with @dineshbhor this is a bug. On the other hand the unit test comes handy :)

damemi commented:
@dineshbhor ah sorry, my mistake, I was thinking this was more part of the feature request from https://github.com/kubernetes-sigs/descheduler/issues/725. A unit test will still be good to have though

dineshbhor commented:
Thanks, @ingvagabund @damemi for the comments. I added the unit tests.
Please check when you get time. 🙇 

I also would like to know what is the release policy for descheduler and when the new image can be available? Please direct me to the available related document if possible.

Thank you 🙇

ingvagabund commented:
@dineshbhor Thank You for addressing the comments. Would you mind squashing before the final approval?

dineshbhor commented:
> @dineshbhor Thank You for addressing the comments. Would you mind squashing before the final approval?

Thanks, @ingvagabund for the comment. Done

dineshbhor commented:
> https://github.com/kubernetes-sigs/descheduler/issues/735 for a start

> 01:00 UTC Wednesday 30th March 2022: Week 12 — [Code Freeze](https://github.com/kubernetes-sigs/descheduler/release_phases.md#code-freeze)

Oh, Tomorrow is code freeze. I hope to get this PR included in the next release. 🙇 
Thanks all for the review.

ingvagabund commented:
/approve
/lgtm

damemi commented:
@dineshbhor we don't cut the release for descheduler until after the upstream Kubernetes release, so code freeze isn't a problem, this will make it in to the next descheduler release

dineshbhor commented:
Thanks, @damemi for sharing the information. 🙇
